External ref resolution using authentication configuration by approved Host

rgranadosd opened the pull request:
Proposal to support authentication for external references in OpenAPI by providing a configuration per approved host.

![image](https://github.com/42Crunch/vscode-openapi/assets/16778975/a6d4211e-43c6-4e09-9f39-cb1e34b7c66a)

This pull request primarily introduces changes to improve the handling of external references in OpenAPI files, particularly for authenticated services. The updates include a new configuration for authentication, modifications to the `package.json` file, and changes to the `src/cache.ts` and `src/external-refs.ts` files to support the new authentication configuration. 

Authentication Configuration:

* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R116-R123): Added a new section explaining how to configure authentication for external references in OpenAPI files. This allows users to specify authentication for specific hosts in the Approved Hosts section of the extension configuration.

Changes to `package.json`:

* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L803-R820): The `openapi.approvedHostnames` property has been renamed to `openapi.approvedHosts` and its type has been changed from `array` to `object`. This allows users to specify not just the hostnames but also the corresponding authorization tokens/headers for resolving external HTTP/HTTPS references.

Changes to `src/cache.ts`:

* [`src/cache.ts`](diffhunk://#diff-d027d3b9853e6bedddf63bd1d8fa67c69b4c1ff2b947685108f315e4e831cdbdL7-R7): Imported the `getApprovedHosts` function from `external-refs` and used it to get the list of approved hosts instead of directly accessing the configuration. This change is made in the `bundle` method of the `BundledDocumentCache` class. [[1]](diffhunk://#diff-d027d3b9853e6bedddf63bd1d8fa67c69b4c1ff2b947685108f315e4e831cdbdL7-R7) [[2]](diffhunk://#diff-d027d3b9853e6bedddf63bd1d8fa67c69b4c1ff2b947685108f315e4e831cdbdL236-R236)

Changes to `src/external-refs.ts`:

* [`src/external-refs.ts`](diffhunk://#diff-a5c3c8d75d1186e864c17c4c77a3ae00befdec67af9b6f5619942856507c6e80L7-R7): Imported `OptionsOfTextResponseBody` from `got` and added several new functions to handle the new `approvedHosts` configuration. The `getApprovedHosts` function was also exported for use in other files. [[1]](diffhunk://#diff-a5c3c8d75d1186e864c17c4c77a3ae00befdec67af9b6f5619942856507c6e80L7-R7) [[2]](diffhunk://#diff-a5c3c8d75d1186e864c17c4c77a3ae00befdec67af9b6f5619942856507c6e80R63-R138)
* [`src/external-refs.ts`](diffhunk://#diff-a5c3c8d75d1186e864c17c4c77a3ae00befdec67af9b6f5619942856507c6e80L75-R168): In the `ExternalRefDocumentProvider` class, updated the `provideTextDocumentContent` method to check if a host is approved before fetching content. Also added handling for the authorization header if provided in the host's configuration.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/42Crunch/vscode-openapi?pullRequest=256) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/42Crunch/vscode-openapi?pullRequest=256) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ak1394<br/>:x: rgranadosd<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/42Crunch/vscode-openapi?pullRequest=256) it.</sub>

ak1394 commented:
Hi @rgranadosd Thanks for your PR! 

I'd like to address couple of issues with it before I can merge it: 

1. Authentication information is stored unencrypted in the configuration, instead I'd suggest to use SecureStorage to store it.
2. Changing exisiting `approvedHostnames` configuration item to `approvedHosts` will affect existing users who have configured this setting.

I think the best way to address it, is to keep `approvedHostnames` unchanged, but store additional authentication information in the SecureStorage. 

SecureStorage, unlike the configs,  does not have a UI that would allow user to add/modify information there, so a UI for adding/updating auth data needs to be implemented.

There is somewhat similar UI implemented in "42Crunch: Open Settings" command, for configuring a "Runtime -> Scand manager" authentication, with relevant code: https://github.com/42Crunch/vscode-openapi/blob/master/packages/web/src/app/config/screen/scan-runtime.tsx

rgranadosd commented:
Hi @ak1394 

Thanks for your feedback, we have implemented the proposed changes from your previous comment, including:
- Webview integration for configuring authentication parameters and approved hostnames
- Usage os SecretStorage to safely store authentication data
 
We have also removed the approvedHostnames configuration and implemented a process to automatically adapt the old configuration parameter to the new model, in a way transparent to users.
 
Please give it a look and tell us with any feedback you might have on the new proposal.
 
Best regards

rgranadosd commented:
> Thanks a lot for the updated PR, and sorry for a bit of delay reviewing it!
> 
> It looks like the `screen/openapi-external-refs.tsx` was not included in the recent changes, could you add it please?

Hi  @ak1394 
 
indeed a file were missing from our previous commit, sorry, we have already included this file.
 
Thank you and best regards

spanishkangaroo commented:
Hello @ak1394,
I would love to have this feature also on the IntelliJ extension!
Do you have any plans on porting it to IntelliJ or an estimate date?
Best regards,

spanishkangaroo commented:
Good morning @ak1394,
thanks a lot! The changes you thought about seem great, so do not hesitate to directly suggest, add them, etc.
Best regards,

ak1394 commented:
Hi Guys!

Below the feedback for the PR. I must note that you managed to do quite a complex change with no help from our side! Feel free though to contact me directly if you wanted to discuss any further implementation details (we can do a voice call, screenshare etc., ping me if you want my contact details). I believe  that your PR is useful and adds to the functionality of the extension, so thanks a lot for your effort.

### Keep `approvedHostnames`

I would prefer to keep `approvedHostnames` configuration item as is, I don't think there is a need to deprecate it and deal with `moveOldApprovedHostnamesConfiguration()` especially as we'll have to do the same in IntelliJ (and Eclipse, if you're interested in Eclipse at all :) )

Let's think about it as follows - if the host is present in `approvedHostnames`, it can be used for $ref's and maybe it has some related authentication information in secret storage. If  there is nothing in storage, then it can be accessed unauthenticated.

This way the hosts that require no authentication can be configured the old way, and we can use the config webapp to add authentication info as required.

Also, the existing function `registerAddApprovedHost()` in `src/external-refs.ts` can be kept as is (it would have to change if we remove approvedHostnames).

So, majority of changes will go into `src/util/config.ts` to read the data from config / secrets, merge it into `ApprovedHostConfiguration` objects that would be send to webapp. Format of the data that is sent to webapp can be kept as in the PR (i.e. `approvedHosts` is sent to the webapp, and webapp doesn't have to know about approvedHostnames)


### Update `screen/openapi-external-refs.tsx` to use react-hook-form

I use react hook form for working form inputs in the config screens, instead of dispatching redux actions directly.

Please have a look at `https://github.com/42Crunch/vscode-openapi/tree/master/packages/web/src/app/scanconf/environment` to see how it is used, in particular `EnvironmentForm.tsx` and its use of `useFieldArray()` to render a table.

Bonus point, if you could use CSS Grid as an example instead of tables, just to keep the UI implementation uniform.

### Rebase, squash

Please rebase on the latest master and squash the commits.

rgranadosd commented:
@ak1394 the requested changes were applied, please review it when you can.
 Thanks in advance

ak1394 commented:
@rgranadosd thanks for the changes! will update you once reviewed

ak1394 commented:
Hi @rgranadosd ! I've emailed you with some info regarding this PR.
